### ISSUE TITLE
input_filename is always <stdin>

### ISSUE THREAD

**Issue by qistoph** [kislyuk/yq#84]
I would expect `input_filename` to work like it does in jq, with yq however it always outputs `<stdin>`.

Actual result:
```
$ yq 'input_filename' */problem.yaml
"<stdin>"
"<stdin>"
"<stdin>"
"<stdin>"
"<stdin>"
"<stdin>"
```

Expected result:
```
$ yq 'input_filename' */problem.yaml
"project_alpha/problem.yaml"
"project_zero/problem.yaml"
"questions/problem.yaml"
"research/problem.yaml"
"stuff/problem.yaml"
"test/problem.yaml"
```

**Comment by kislyuk** [comment]
Thanks for pointing this out. Although the value of `input_filename` can be set externally in jq, I'm not sure how one would do that with multiple input documents. Open to suggestions.

**Comment by qistoph** [comment]
Just took a look at your code.
https://github.com/kislyuk/yq/blob/master/yq/__init__.py#L172

It seems to me you are concatenating all input files into one big stream. My guess would be that jq would need you to supply multiple separate input stream to make input_filename work. Don't know exactly how, but I'll try to figure out some more.

**Comment by kislyuk** [comment]
Right. Literally supplying multiple input streams is technically possible (for example, via a bunch of temporary FIFOs), but cumbersome and could be less performant.

**Comment by vergenzt** [comment]
> Although the value of `input_filename` can be set externally in jq, I'm not sure how one would do that with multiple input documents. Open to suggestions.

@kislyuk how would you set the value of `jq`'s `input_filename` externally without passing the path to a real file or pipe via command line arguments? I've spelunked in the `jq` codebase for a bit but can't figure out how one would cause `input_filename` to be a string other than one that was passed via command line args (either explicitly or through a glob).

**Comment by kislyuk** [comment]
@vergenzt you can set the `input_filename` variable via a jq command line argument.

**Comment by vergenzt** [comment]
@kislyuk which command line option are you seeing that does that? I don't see anything mentioned about this in the [documentation](https://stedolan.github.io/jq/manual/v1.6/#Invokingjq).

And as far as I could tell in the codebase, the value returned by [`input_filename`][input_filename] function is the same as the value passed to [`fopen`][fopen] (with the exception of setting the filename to `"<stdin>"` if input == stdin). There's `-f` / `--from-file`, but that just reads the filter from that file, it doesn't set `input_filename`.

Code references:
 - https://github.com/stedolan/jq/blob/jq-1.6/src/builtin.c#L1653
 - https://github.com/stedolan/jq/blob/jq-1.6/src/builtin.c#L1550-L1558
 - https://github.com/stedolan/jq/blob/jq-1.6/src/util.c#L393
 - https://github.com/stedolan/jq/blob/jq-1.6/src/util.c#L291-L305

[input_filename]: https://stedolan.github.io/jq/manual/v1.6/#input_filename
[fopen]: http://man7.org/linux/man-pages/man3/fopen.3.html

**Comment by vergenzt** [comment]
My current understanding is that the only way to control `input_filename` is by actually creating a file on the filesystem with the given path/name.

Would you be open to a PR adding some sort of `--use-temp-fifos` option to create temporary named FIFOs for each input file? Could use `mktemp` with a format of `<filename>.temp_XXXXX.json` so that the original filename is recoverable by `input_filename | sub(".temp_\w+.json$", "")`.

(Or personally I'd be a fan of just _testing_ whether there's a noticeable performance hit with a large number of files, and if there's not then just defaulting to making the FIFOs.)

E.g.  for original files `./myfile1.yaml` and `./myfile2.yaml`, create FIFOs `./myfile.yaml.temp_In49e.json` and `./myfile2.yaml.temp_coHi9.json` and pass those to `jq` instead of concatenating them on stdin.

**Comment by kislyuk** [comment]
@vergenzt you're right, there is no way to set `input_filename` - in the test I ran earlier, I confused `input_filename` with `$input_filename`.

**Comment by qistoph** [comment]
How about checking with the jq authors to see if they're willing in include an argument of variable to enable this functionality?

**Comment by kislyuk** [comment]
@qistoph please feel free to do that. However, I don't think it will solve this issue unless you also want to design a protocol for naming all documents in a stream and passing the names of those documents as a list to jq.

**Comment by qistoph** [comment]
Getting the filename is nowadays possible with `filename`.

```
yq 'filename' */docker-compose.yaml
```

**Comment by qistoph** [comment]
```
$ yq --version
yq (https://github.com/mikefarah/yq/) version v4.33.2
$ yq '{(filename): .services|length}'
 */docker-compose.yaml
InfluxForwarder/docker-compose.yaml: 1
NetMonitor/docker-compose.yaml: 1
authentik/docker-compose.yaml: 4
download/docker-compose.yaml: 5
fritzcollect/docker-compose.yaml: 1
heimdall/docker-compose.yaml: 1
influxdb/docker-compose.yaml: 1
jellyfin/docker-compose.yaml: 1
jupyter/docker-compose.yaml: 1
mariadb/docker-compose.yaml: 2 
```

**Comment by vergenzt** [comment]
@qistoph that's the wrong `yq`. 🙃  https://github.com/mikefarah/yq/ is a different YAML CLI helper that's doesn't use `jq` and is not compatible.

**Comment by stagnation** [comment]
I too would prefer to have the `input_filename`. Though it was possible to work around it in my xargs pipeline by setting the external argument `--arg <yamlfile> {}` and then accessing it in the program instead of the regular `input_filename`

```find . -name ... | xargs -I{} yq --arg yamlfile {} '.tasks[] | { label: label, file: $yamlfile, ... }```
